Integrations/MicrosoftGraphGroups/MicrosoftGraphGroups.py

```
@@ -0,0 +1,441 @@
+                     next_link: str = None):
+        """
+        Generic request to Microsoft Graph
+        """
+        token = self.get_access_token()
+        if next_link:
+            url = next_link
+        else:
+            url = f'{self.base_url}{url_suffix}'
+
+        response = requests.request(
+            method,
+            url,
+            headers={
+                'Authorization': 'Bearer ' + token,
+                'Content-Type': 'application/json',
+                'Accept': 'application/json'
+            },
+            params=params,
+            data=body,
+            verify=self.use_ssl,
+        )
+        try:
+            data = response.json() if response.text else {}
+            if not response.ok:
+                return_error(f'API call to MS Graph failed [{response.status_code}]'
+                             f' - {demisto.get(data, "error.message")}')
+            elif response.status_code == 206:  # 206 indicates Partial Content, reason will be in the warning header
+                demisto.debug(str(response.headers))
+
+            return data
+
+        except TypeError as ex:
+            demisto.debug(str(ex))
+            return_error(f'Error in API call to Microsoft Graph, could not parse result [{response.status_code}]')
+
+    def test_function(self):
+        token = self.get_access_token()
+        response = requests.get(
+            self.base_url + 'groups',
+            headers={
+                'Authorization': 'Bearer ' + token,
+                'Content-Type': 'application/json',
+                'Accept': 'application/json'
+            },
+            params={'$orderby': 'displayName'},
+            verify=self.use_ssl
+        )
+        try:
+            data = response.json() if response.text else {}
+            if not response.ok:
+                return_error(f'API call to MS Graph failed. Please check authentication related parameters.'
```

you should raise formatted exception instead of return error. this should be handled in the main func.